### PR TITLE
Dagster x SDF Integration: Introducing `sdf_assets` decorator and `SdfWorkspace`

### PR DESCRIPTION
**Pull request by akbog**
## Summary & Motivation

Hurray 🎉 This is the first PR of many towards the goal of enabling the orchestration of SDF Workspaces in Dagster.

The version of the dagster-sdf integration outlined in this PR is able to handle the `moms_flower_shop` sdf workspace, including metadata extraction and local materialization via the sdf db. We introduce the following concepts to accomplish this:

1. `sdf_assets` decorator

Wraps a `multi_asset` definition, which simply defines the workspace as a single op / graph. This can make sense in the context of an sdf workspace for several reasons (namely caching). The decorator takes as inputs:

  - workspace_dir: the root of an sdf workspace (i.e. the file with a `workspace.sdf.yml` in it)
  - target_dir: the target path to store sdf outputs in (i.e. materialized data from the sdf db, cache, etc.)
  - environment: which environment to run sdf commands against (default: dbg)

The decorator namely depends on the `SdfInformationSchema`.

2. `SdfInformationSchema` class

This is a wrapper class for interacting with sdf's compile time generated information schema. The information schema is a store of various metadata, such as classifiers, table materializations (i.e. table, view, external etc.), table and column level lineage, and table schema information. We load the information schema using pandas, extracting the dagster dag and schema information from the tables and columns dataframes respectively. These are both located in the sdf_target dir under `sdf_target/data/system/information_schema::sdf/`.

3. `SdfWorkspace` class

Wraps an sdf workspace and workspace_preparer. The `SdfWorkspacePreparer` is used to prepare an sdf workspace (i.e. run compile) if `DAGSTER_SDF_COMPILE_ON_LOAD=1` when starting the dagster web server. This ensures that the project dag is available. It also ensures that metadata, such as Table Schema information, table and column level classifiers, and eventually column level lineage, is available pre-materialization.

4. `SdfCliResource` class

Responsible for making calls to the sdf cli binary, and interpreting the results through log processing. We invoke the  sdf cli with `--log-level=info` and `--log-form=nested` which results in structured logs that tell dagster when a table has finished materializing. We do not extract table schema, or other metadata from structured logs, as this is all available at compile time via the information schema.

**Please note**, this work is heavily-inspired by the work of @rexledesma so fair credit to him and the development team at dagster!

## How I Tested These Changes

There are new tests that run off of checked-in sdf workspaces. See [dagster_sdf_tests](dagster_sdf_tests). These ensure proper workspace initialization, information schema ingestion, and dag generation. More tests to follow in subsequent PRs.

## More to come

1. DagsterSdfTranslator Class
2. dagster-sdf project scaffold

### CODE REVIEW

File: python_modules/libraries/dagster-sdf/dagster_sdf/sdf_cli_resource.py [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+
+SdfDagsterEventType = Union[Output, AssetMaterialization, AssetCheckResult, AssetObservation]
+
+@dataclass
+class SdfCliInvocation:
+    """The representation of an invoked sdf command.
+
+    Args:
+        process (subprocess.Popen): The process running the sdf command.
+        information_schema (SdfInformationSchema): The information schema for sdf nodes.
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/sdf_cli_resource.py:151]: This does not actually exist in the list of arguments. Should we remove?

**Review by rexledesma — CHANGES_REQUESTED** [review]
Thanks for splitting this up from the initial PR. However, there's still a lot of moving parts here (68 files, >1.5k lines changed, test coverage is still minimal).

We cannot merge this until we split this up even further so that the changes are more tractable.

Here's a proposed plan of approach to decompose this problem further:

1. Templates of the sdf projects (e.g. `moms_flower_shop`, `csv_123`) should be committed in separate PRs.
2. Implement `SdfCliResource` and add tests in a separate PR.
3. Implement `SdfInformationSchema` and add tests in a separate PR.
4. Implement `SdfWorkspace` and add tests in a separate PR.
5. Implement `sdf_assets` and add tests in a separate PR.

File: python_modules/libraries/dagster-sdf/setup.py [hunk 2]
```diff
@@ -34,7 +38,14 @@ def get_version() -> str:
     ],
     packages=find_packages(exclude=["dagster_sdf_tests*"]),
     python_requires=">=3.8,<3.13",
-    install_requires=[f"dagster{pin}"],
+    install_requires=[
+        f"dagster{pin}",
+        f"sdf-cli>=0.3.8,<{SDF_VERSION_UPPER_BOUND}",
+        "orjson",
```
[the remark this answers is not in the document]
Comment (akbog) [reply python_modules/libraries/dagster-sdf/setup.py:44]: This is used when reading in raw events from the cli's structured logs

File: python_modules/libraries/dagster-sdf/setup.py [hunk 3]
```diff
@@ -34,7 +38,14 @@ def get_version() -> str:
...
+        "pandas",
+        "pyarrow",
```
[the remark this answers is not in the document]
Comment (akbog) [reply python_modules/libraries/dagster-sdf/setup.py:46]: Parquet is currently the only supported output format for the information schema! We can explore some more light weight dataframe libraries once this initial push is in but I totally hear you

File: python_modules/libraries/dagster-sdf/dagster_sdf_tests/sdf_workspaces/__init__.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+from pathlib import Path
+
+projects_path = Path(__file__).joinpath("..").resolve()
+
+test_csv_123_path = projects_path.joinpath("csv_123")
+moms_flower_shop_path = projects_path.joinpath("moms_flower_shop")
```
[the remark this answers is not in the document]
Comment (akbog) [reply python_modules/libraries/dagster-sdf/dagster_sdf_tests/sdf_workspaces/__init__.py:6]: These are sample projects, same as jaffle_shop. See our docs for a moms_flower_shop tutorial: https://docs.sdf.com/tutorials/tutorials-intro#getting-started-with-moms-flower-shop

File: python_modules/libraries/dagster-sdf/dagster_sdf/errors.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+from dagster import (
+    Failure
+)
+
+
+class DagsterSdfError(Failure, ABC):
+    """The base exception of the ``dagster-sdf`` library."""
+
+class DagsterSdfCliRuntimeError(DagsterSdfError, ABC):
```
[the remark this answers is not in the document]
Comment (akbog) [reply python_modules/libraries/dagster-sdf/dagster_sdf/errors.py:13]: It is used by the `sdf_cli_resource`

[the remark this answers is not in the document]
Comment (rexledesma) [reply python_modules/libraries/dagster-sdf/dagster_sdf/errors.py:13]: Ah you're right. It didn't show since that file was too large to render by default

**Pull request closed without merging**